mlabs/src/Mlabs/NFT/Contract.hs

```diff
@@ -0,0 +1,375 @@
+import Prelude (mconcat, (<>))
+import Prelude qualified as Hask
+
+import Control.Lens (filtered, to, traversed, (^.), (^..), _Just, _Right)
+import Control.Monad (join, void)
+import Data.List qualified as L
+import Data.Map qualified as Map
+import Data.Monoid (Last (..))
+import Data.Text (Text, pack)
+
+import Text.Printf (printf)
+
+import Plutus.Contract (Contract, Endpoint, endpoint, utxosTxOutTxAt, type (.\/))
+import Plutus.Contract qualified as Contract
+import PlutusTx qualified
+
+import Ledger (
+  Address,
+  ChainIndexTxOut,
+  Datum (..),
+  Redeemer (..),
+  TxOutRef,
+  ciTxOutDatum,
+  ciTxOutValue,
+  getDatum,
+  pubKeyAddress,
+  pubKeyHash,
+  scriptCurrencySymbol,
+  txId,
+ )
+
+import Ledger.Constraints qualified as Constraints
+import Ledger.Typed.Scripts (validatorScript)
+import Ledger.Value as Value (TokenName (..), singleton, unAssetClass, valueOf)
+
+import Playground.Contract (mkSchemaDefinitions)
+
+import Mlabs.NFT.Types (
+  BuyRequestUser (..),
+  Content (..),
+  MintParams (..),
+  NftId (..),
+  QueryResponse (..),
+  SetPriceParams (..),
+  UserId (..),
+ )
+
+import Mlabs.NFT.Validation (
+  DatumNft (..),
+  NftTrade,
+  UserAct (..),
+  asRedeemer,
+  calculateShares,
+  mintPolicy,
+  nftAsset,
+  nftCurrency,
+  priceNotNegative,
+  txPolicy,
+  txScrAddress,
+ )
+
+import Mlabs.Plutus.Contract (readDatum', selectForever)
+
+-- | A contract used exclusively for query actions.
+type QueryContract a = Contract QueryResponse NFTAppSchema Text a
+
+-- | A contract used for all user actions.
+type UserContract a = Contract (Last NftId) NFTAppSchema Text a
+
+-- | A common App schema works for now.
+type NFTAppSchema =
+  -- Author Endpoint
+  Endpoint "mint" MintParams
+    -- User Action Endpoints
+    .\/ Endpoint "buy" BuyRequestUser
+    .\/ Endpoint "set-price" SetPriceParams
+    -- Query Endpoints
+    .\/ Endpoint "query-current-owner" NftId
+    .\/ Endpoint "query-current-price" NftId
+
+mkSchemaDefinitions ''NFTAppSchema
+
+-- MINT --
+
+-- | Mints an NFT and sends it to the App Address.
+mint :: MintParams -> UserContract ()
+mint nftContent = do
+  addr <- getUserAddr
+  nft <- nftInit nftContent
+  utxos <- Contract.utxosAt addr
+  oref <- fstUtxo addr
+  let nftId = dNft'id nft
+      scrAddress = txScrAddress
+      nftPolicy = mintPolicy scrAddress oref nftId
+      val = Value.singleton (scriptCurrencySymbol nftPolicy) (nftId'token nftId) 1
+      (lookups, tx) =
+        ( mconcat
+            [ Constraints.unspentOutputs utxos
+            , Constraints.mintingPolicy nftPolicy
+            , Constraints.typedValidatorLookups txPolicy
+            ]
+        , mconcat
+            [ Constraints.mustMintValue val
+            , Constraints.mustSpendPubKeyOutput oref
+            , Constraints.mustPayToTheScript nft val
+            ]
+        )
+  void $ Contract.submitTxConstraintsWith @NftTrade lookups tx
+  Contract.tell . Last . Just $ nftId
```

Comment:
@cstml why is there "Just" here?

Reply:
I think the `Last . Just` could be replaced with `return`, if that's a better way.